[DeviceMesh] Add `_unflatten_` api for device mesh to support better UX for some use cases like EP and replicate

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159482

After some initial feedback on the implementation of `_split`, we realize that we can first implement `_unflatten` for urgent use cases ask for now. And we will do more refactoring and iterations based on the discussions from this PR and this RFC: https://github.com/pytorch/pytorch/issues/159013. We will also ensure that all changes won't cause regression to DTensor's CPU overhead as well.


This PR is trying to:
1. For unflatten, we don't support flatten a unflattened mesh which might not be necessary because by the time when user decide to flatten a unflatten, users might essentially redo the unflatten operations which could make bookeeping complicated to handle and we don't see that use cases for now. (We throw a NotImplementError for now)
3. We need some extra book-keeping for unflatten api, what we want to keep track is which sub-mesh contains the unflattened `dim_name` so that users can slice these dim_names from root mesh as well. And we need to swap the mesh to slice when slicing from root mesh. Also to make sure we don't unflatten same `dim_name` into different sizes, we need to keep the total accumulated numel in the root for that dim_name as well.
4. We want to reuse PGs already created for the same `dim_name`. For the case when a different `dim_name` happens to have different shapes, we will create new PG because with a different name, users might want to use that dimension for different purposes, so we'd better not to reuse. (This assumption can be changed, so I am open to suggestions)
5. Added unit test to two situation: 1. we directly do unflatten on one 2D device mesh. 2. we first create a dummy 1D device mesh and then split into two 3D device mesh.

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta

fduwjj committed:
[WIP][DeviceMesh] Add _split api for device mesh

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP][DeviceMesh] Add _split api for device mesh"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP][DeviceMesh] Add _split api for device mesh"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP][DeviceMesh] Add _split api for device mesh"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP][DeviceMesh] Add `_split` api for device mesh"


We decide to implement split based on our initial discussion in https://github.com/pytorch/pytorch/issues/159013. This PR is trying to:
1. Expose `_init_backend` in init_device_mesh so that users can create a device mesh without any backend as a default world.
2. For split, we don't support flatten a split mesh which might not be necessary since if user decide to split, users might start to split from a flattened dim already. (There are definitely corner cases, so I throw a NotImplementError)
3. We need some book-keeping for split api, we thing we want keep track is what sub-mesh contains the split dim_name so that users can slice these dim_names from root mesh as well. And we need to swap the mesh to slice when slicing from root mesh. Also to make sure we don't split same dim_name into different sizes, we need to keep the total accumulated numel for that dim_name as well.
4. We want to reuse PGs already created for the same dim_name. For different dim_name happens to have different shapes, we will create new PG because with a different name, users might want to use that dimension for different purposes, so we'd better not to reuse. (This assumption can be changed, so I am open to suggestions)
5. Added unit test to two situation: 1. we directly do split on one 2D device mesh. 2. we first create a dummy 1D device mesh and then split into two 3D device mesh.

There are definitely rough edges, so I just want to send out this PR first and gather early feedback to see if this is a reasonable direction or not.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k 

[ghstack-poisoned]

wanchaol reviewed (commented):
Some suggestions on the API

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -988,11 +1173,36 @@ def _flatten(self, mesh_dim_name: Optional[str] = None) -> "DeviceMesh":
         mesh_shape: tuple[int, ...],
         *,
         mesh_dim_names: Optional[tuple[str, ...]] = None,
+        _init_backend: bool = True,
```
I think for this case, we don't even want to create PG at all. But you are right, current c10d relies on a default World PG anyway so this is a no-op..

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -988,11 +1173,36 @@ def _flatten(self, mesh_dim_name: Optional[str] = None) -> "DeviceMesh":
             return _mesh_resources.create_flatten_mesh(self, mesh_dim_name)
 
+        def _split(
+            self, split_sizes: tuple[int, ...], mesh_dim_names: tuple[str, ...]
```
I am ok with the name and if it conjugates with the current API. Question about EP's UX, say we have this `_unflatten`

Does the UX like the blow?
```
# Create a dummy 1D device mesh first, so no PG will be created under the hood. This is important for large-scale because no global PG will be used in that case.
device_mesh = init_device_mesh(
    'cuda', (32,), mesh_dim_names=["world"],
)
device_mesh._unflatten(0, (4,2,4), mesh_dim_names=["dp", "cp", "tp"])
device_mesh._unflatten(0, (2,8,2), mesh_dim_names=["dp_ep", "ep", "tp_ep"])
device_mesh["cp", "tp"]["tp"]
```

Also, aside from the API's name and UX, do you have any comments on the bookeeping and other changes to DeviceMesh, because now with unflatten, we need to map from root mesh to the unflattened mesh. Do they look reasonable as well?

fduwjj committed:
Update on "[WIP][DeviceMesh] Add `_split` api for device mesh"


We decide to implement split based on our initial discussion in https://github.com/pytorch/pytorch/issues/159013. This PR is trying to:
1. Expose `_init_backend` in init_device_mesh so that users can create a device mesh without any backend as a default world.
2. For split, we don't support flatten a split mesh which might not be necessary since if user decide to split, users might start to split from a flattened dim already. (There are definitely corner cases, so I throw a NotImplementError)
3. We need some book-keeping for split api, we thing we want keep track is what sub-mesh contains the split dim_name so that users can slice these dim_names from root mesh as well. And we need to swap the mesh to slice when slicing from root mesh. Also to make sure we don't split same dim_name into different sizes, we need to keep the total accumulated numel for that dim_name as well.
4. We want to reuse PGs already created for the same dim_name. For different dim_name happens to have different shapes, we will create new PG because with a different name, users might want to use that dimension for different purposes, so we'd better not to reuse. (This assumption can be changed, so I am open to suggestions)
5. Added unit test to two situation: 1. we directly do split on one 2D device mesh. 2. we first create a dummy 1D device mesh and then split into two 3D device mesh.

There are definitely rough edges, so I just want to send out this PR first and gather early feedback to see if this is a reasonable direction or not.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj commented:
Discussed with @lw offline, we might need to do more refactors down the road but this PR is aiming at unblocking some urgent use cases from EP, so as long as we make UX and API signature correct, we will perform clean up later. And this is going to be multiple steps and we will make sure no regression to DTensor's overhead as well.

fduwjj commented:
Updated the PR based on the feedback from @wanchaol and I will do more cleanups to the logic of bookeepings part so it look cleaner. All UT passed. We also need to rebase on top of https://github.com/pytorch/pytorch/pull/159371.

fduwjj committed:
Update on "[WIP][DeviceMesh] Add `_unflatten_` api for device mesh"


After some initial feedback on the implementation of `_split`, we realize that we can first implement `_unflatten` for urgent use cases ask for now. And we will do more refactoring and iterations based on the discussions from this PR and this RFC: https://github.com/pytorch/pytorch/issues/159013. We will also ensure that all changes won't cause regression to DTensor's CPU overhead as well.


This PR is trying to:
1. For unflatten, we don't support flatten a unflattened mesh which might not be necessary because by the time when user decide to flatten a unflatten, users might essentially redo the unflatten operations which could make bookeeping complicated to handle and we don't see that use cases for now. (We throw a NotImplementError for now)
3. We need some extra book-keeping for unflatten api, what we want to keep track is which sub-mesh contains the unflattened `dim_name` so that users can slice these dim_names from root mesh as well. And we need to swap the mesh to slice when slicing from root mesh. Also to make sure we don't unflatten same `dim_name` into different sizes, we need to keep the total accumulated numel in the root for that dim_name as well.
4. We want to reuse PGs already created for the same `dim_name`. For the case when a different `dim_name` happens to have different shapes, we will create new PG because with a different name, users might want to use that dimension for different purposes, so we'd better not to reuse. (This assumption can be changed, so I am open to suggestions)
5. Added unit test to two situation: 1. we directly do unflatten on one 2D device mesh. 2. we first create a dummy 1D device mesh and then split into two 3D device mesh.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

wanchaol reviewed (commented):
Trying to review this PR in details but the unflatten logic feels pretty complicated to understand. I think we should try to simplify the implementation, i.e.
* Make fewest assumptions, i.e. let's not get into the business of "same unflatten shape + different unflatten mesh dim names", unless there's a clear use case for it. I think simply error out in that case make more sense. 
* For unflatten I don't quite feel the "implicit mesh dim appear in the root mesh" is a good UX, it's quite surprising behavior IMO and not quite necessary. So I suggest to not make it part of feature of `unflatten` and later consider deprecate this behavior in `flatten`. 

I think for this PR, let's only make the essential feature to work, this would help simplify the implementation.

wanchaol commented:
> However, I'm wondering if we can rework DeviceMesh slightly to simplify all this and handle all corner cases cleanly. Concretely, a "full" device mesh is a tensor, thus I'm tempted to reuse the same separation as tensor-vs-storage also for DeviceMeshes. What this means is that each slices/flattened/unflattened "derived" DeviceMesh is basically a view into the "root" original DeviceMesh (i.e., the storage), and such a view can likely be expressed as an offset, a (sub-)size and a stride (or something like this, I'm sure we can figure it out).
> 
> Thus we can define a single generic function that takes this offset/shape/stride combination and creates a new sub-PG for it. This function can then be used both by the `_flatten` codepath and by the `_unflatten`/`_split` one. Then, on the root DeviceMesh, we can keep a dictionary indexed by this offset/shape/stride tuple, so that we can detect whether a compatible PG for a certain flatten/split had already been created in the past and can be reused. And, finally, we can add a map between "names" (for dimensions) and tuples of offset/shape/stride so that we can avoid introducing ambiguous duplicate names and we can retrieve the PG for a named dimension.
> 

@lw I think overall this make sense to me! Probably need to get into the details to figure out how exactly the tensor <-> storage mapping would work, but yeah as long as it could simplify the existing implementation, I'm all for it :)

wanchaol commented on test/distributed/test_device_mesh.py:
```diff
@@ -883,6 +883,44 @@ def test_flatten_mesh_4d(self):
+        self.assertEqual(
+            unflatten_mesh.mesh_dim_names, ["dp_shard", "dp_replicate", "tp"]
+        )
+        self.assertEqual(mesh_2d["tp"].mesh, unflatten_mesh["tp"].mesh)
```
The unflatten API itself is the correct UX I agreed on, but the way to access the unflattened device meshes from the root mesh is not, so basically the last line is the discrepency: `device_mesh["cp", "tp"]["tp"]`, instead user should use the unflattened result to access the device mesh I think. 

The reason I go against this behavior is that accessing from the root mesh is quite an implicit behavior, this is not really my design principle, but the Python/PyTorch [design principle](https://docs.pytorch.org/docs/stable/community/design.html#principle-2-simple-over-easy): Explicit is better than implicit!

To answer your question, when the `flatten` API was designed, I was also NOT a fan of this behavior, but given `flatten` only add one dimension to the root mesh, it did not quite contaminate this principle. Concerns about this was also the main reason I wanted to keep `flatten` private first. But now with the design of this `unflatten` API, I think the contamination become much worse. Why? Ideally user shouldn't aware of the root mesh concept as that is mainly an implementation detail, so the "root mesh" and the unflattened mesh should simply both be device meshes that represents a n-d array of devices. But now with the `unflatten` + root mesh behavior, a 1-D root device mesh suddenly have a bunch of additional dimensions! Taking the example you gave:

* `device_mesh` is a 1-D device mesh, but suddenly after the `unflatten`, it got **6 more** dimensions! 6-more dimensions on a 1-D device mesh does not make sense as an 1-D device array abstraction
* What's worse is that it might not only have 6, it could have 8, 10 or even more dimensions because of calling the `unflatten` twice or more!

The implicit addition of device mesh dimensions make the abstraction not clean, and IMO we should avoid this implicit behavior.

fduwjj replied:
I see, thanks for sharing your thought finally. I do agree that explicit is better than implicit so it makes sense that ideally we can have a more functional-like behavior. Since we already have the principle, we should have rejected the contamination earlier and I was under the impression that we did have a long discussion around `flatten` api as well. But somehow we still end up having a somehow contaminated api which is kind of suboptimal in the first place. OK let me also ask users like @tianyu-l 's input on this as well.

wanchaol replied:
@fduwjj Wow!
From your comments, it seems you’re focusing on assigning personal blame rather than truly wanting to solve the problem and improving PyTorch. It's unclear whether you’re blaming those who discussed `flatten` — such as me, @wz337 and @fegin, or just me. I’m surprised to see this level of unprofessionalism from someone contributing to the PyTorch OSS project. Sarcasm could not help the design or make the APIs be better.

If you are so much obsessed with the "suboptimal stories in the first place". I also do not mind to talk publicly about the story of the TP APIs :)

wanchaol replied:
WOW

@fduwjj From your comments it’s hard to miss that sarcasm beats actually solving the problem or improving PyTorch. If the blame is meant for those who discussed `flatten`—me, @wz337, @fegin—or just me, please say so. I’m sorry to see this level of unprofessionalism from someone contributing to the PyTorch OSS project. Sarcasm won’t make the design or the APIs better.

If you do need help, folks here are happy to chime in—we’d really like to avoid another TP API situation :)

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -238,6 +250,186 @@ def create_flatten_mesh(
+            # We need to check whether one dim_name is used in unflatten into more than one shape/dim.
+            # For example, if we have already unflatten into a 3D mesh with mesh_shape (2, 2, 2) mesh_dim_names ("dp, "cp", "tp")
+            # and we want to unflatten into a 2D mesh with mesh_shape (4, 2) mesh_dim_names ("dp", "cp"), this is wrong.
+            # But unflatten into another 3D mesh with mesh_shape (2, 2, 2) mesh_dim_names ("dp, "ep", "ep_tp") is legit.
```
I see, based on your comment above, since we like to hold the principle of being explicit, I agree we can just reuse it to make the behavior simple and predictable.
